implement C++23 `byteswap`

davebayer opened the pull request:
This PR introduces C++23 `std::byteswap` to CCCL and makes it available back in C++11.

The implementation uses compiler intrinsics `__builtin_bswap` if available.

miscco commented on libcudacxx/include/cuda/std/__cccl/builtin.h:
```diff
@@ -126,6 +126,38 @@
+#  define _CCCL_BUILTIN_BSWAP16(...) __builtin_bswap16(__VA_ARGS__)
+#endif // _CCCL_CHECK_BUILTIN(builtin_bswap16)
+
+#if !_CCCL_COMPILER(GCC) || !_CCCL_COMPILER(CLANG) || !_CCCL_CUDA_COMPILER(CLANG)
+#  undef _CCCL_BUILTIN_BSWAP16
+#endif // !_CCCL_COMPILER(GCC) || !_CCCL_COMPILER(CLANG) || !_CCCL_CUDA_COMPILER(CLANG)
```
Also we want a comment on why this requires the fallback implementation, see e.g the `__builtin_log` cases

fbusato commented:
perfect. I added a set of optimization for CUDA a while ago, and it also includes `byteswap` https://github.com/NVIDIA/cccl/issues/2239

miscco commented on libcudacxx/include/cuda/std/__bit/byteswap.h:
```diff
@@ -0,0 +1,121 @@
+                 _CCCL_AND(!has_single_bit(sizeof(_Integer))))
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI _CCCL_CONSTEXPR_CXX14 _Integer byteswap(_Integer __val) noexcept
+{
+  return static_cast<_Integer>(__byteswap_impl::__impl(__to_unsigned_like(__val)));
```
Sorry I missed it, but we need to fully qualify the function call here
```suggestion
  return static_cast<_Integer>(__byteswap_impl::__impl(_CUDA_VSTD::__to_unsigned_like(__val)));
```

davebayer commented:
> perfect. I added a set of optimization for CUDA a while ago, and it also includes `byteswap` #2239

Thank you for the hint. I've implemented the optimized versions of 32-bit and 64-bit byte swap. The PTX output is more or less identical to the clang-cuda builtin functions now.

miscco commented on libcudacxx/include/cuda/std/__bit/byteswap.h:
```diff
@@ -0,0 +1,173 @@
+
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr uint16_t __impl(uint16_t __val) noexcept
+  {
+#if defined(_CCCL_BUILTIN_BSWAP16)
+    return _CCCL_BUILTIN_BSWAP16(__val);
+#  if _CCCL_STD_VER >= 2014
+    if (!__cccl_default_is_constant_evaluated())
+    {
+#    if _CCCL_COMPILER(MSVC)
+      NV_IF_TARGET(NV_IS_HOST, return _byteswap_ushort(__val);)
+#    endif // _CCCL_COMPILER(MSVC)
+    }
+#  endif // _CCCL_STD_VER >= 2014
+#else
+    return (__val << CHAR_BIT) | (__val >> CHAR_BIT);
+#endif
```
This does not look correct.

MSVC does not support the builtin, so that conditional should be in the second branch
```suggestion
#if defined(_CCCL_BUILTIN_BSWAP16)
    return _CCCL_BUILTIN_BSWAP16(__val);
#else // ^^^ _CCCL_BUILTIN_BSWAP16 ^^^ / vvv !_CCCL_BUILTIN_BSWAP16 vvv
#  if _CCCL_COMPILER(MSVC)
    if (!_CUDA_VSTD::is_constant_evaluated())
    {
      NV_IF_TARGET(NV_IS_HOST, return _byteswap_ushort(__val);)
    }
#  endif // _CCCL_COMPILER(MSVC)
    return (__val << CHAR_BIT) | (__val >> CHAR_BIT);
#endif // !_CCCL_BUILTIN_BSWAP16
```

miscco replied:
This might need `!__cccl_default_is_constant_evaluated`

miscco commented on libcudacxx/include/cuda/std/__bit/byteswap.h:
```diff
@@ -0,0 +1,173 @@
+  }
+
+#if !defined(_LIBCUDACXX_HAS_NO_INT128)
+  _CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI static constexpr unsigned __int128 __impl(unsigned __int128 __val) noexcept
```
We are conventionally using `__uint128_t` in libcu++

miscco commented:
/ok to test

miscco commented:
/ok to test

miscco commented:
/ok to test

miscco commented:
/ok to test
